### DISCUSSION TITLE
Dynamic recomputation of policies

### DISCUSSION THREAD

**Discussion by ronenkat** [fybrik/fybrik#597] [Developers]
Mesh for Data blueprint are re-computed on changes to m4dapplication CR.
However, changes to policies do not affect the CR.
Can we add to the **spec** in the CR an entry such as _"policies-updated-at: <datetime>"_ which would trigger a reconsilder run that will re-compute the policies.
@shlomitk1 @roee88 your views?

**Comment by roee88** [comment]
Background k8s info although you probably know it all already:
1. A reconcile is triggered on changes to controlled CR (typically to metadata or spec), owned CRs (typically to status), arbitrary watched CRs. It can also be triggered directly on an external event although it's rarely used/needed.
1. A reconcile is also scheduled after a reconcile that errored (exponential backoff), a reconcile that returned with Requeue (immediately), a reconcile that returned with RequeueAfter (e.g. to implement periodic reconcile). 
1.  A CR has a `generation` field that is auto incremented on changes to its spec. A common pattern that we also use is to also track the obeserved generation (the last generation we reconciled against). Within a reconcile loop you can't know what triggered it but the generation fields help to differentiate between some cases to some degree. 

Now to the issue you raised and ignoring requirements such as impact observability, we have several options:
1. Field in spec as suggested
2. Using an annotation (instead of spec field)
3. Periodically trigger reconcile to account for policy changes using RequeueAfter

In the laso two cases the generation won't change so we also need to keep track of the last time we checked policies and enable this code path if interval elapsed / annotation timestamp is newer.

The first is a bit misleading IMO (this field does not reflect desired state from the workload perspective) and may cause complexity with respect to handling of error paths because it's unclear what to do if a reconcile fails after only this field changed (which you can't differentiate from any other spec change). 

However, if we do consider impact observability then we need to consider things like giving a grace period to existing applications. A recent discussion on it can be found in https://github.com/IBM/the-mesh-for-data/pull/586#discussion_r633060280. My personal preference would be to react to policy changes without a grace period and without considering requirements that should be fulfilled in a policy manager right now. I see it as an improvement to the current behavior of not reacting at all. Instead, making the changes to support the above while defining a new epic to design and develop an interface that a policy manager can use to fulfill the impact observability requirement would be my suggestion.

**Reply by ronenkat** [reply]
Re Grace period and when to apply policy changes - agree, This should be part of the polices mechanism, not the manager code.
What would be the suggested mechanism?

**Reply by shlomitk1** [reply]
Agree with periodical reconcile of M4DApplications. If no changes have been applied to the spec, the reconcile will check:
1) changes to policies
2) changes to plotter resource
I would prefer an easier way to check whether the policy have been changed rather than evaluate all from scratch.

**Comment by simanadler** [comment]
@ronenkat @shlomitk1 @roee88 This same discussion is taking place with @rachit1arora.  Beyond the technical aspects of how to know when a policy has changed, there are a lot of user experience aspects that need to be taken into account.  In my opinion, when a policy is changed the governance officer needs to indicate for example whether it should affect existing workloads and their currently used data sets or not.  Is it a long running workload, short term, etc?  How is the data user made aware of the change, and how/whether it affect his data?

It's important to capture this issue, as you have done ... but since the policy managers we currently support do not address this issue (yet) then I don't think it's a high priority for us to address it in M4D.  I.e. If M4D would pick up changes to policies and implement them without there being any mechanism for making the data users aware of the change nor the governance officers aware that they are affecting data users, then we may be creating a problem.

**Reply by ronenkat** [reply]
@simanadler you mention multiple topics.
1. What is the m4d mechanism to handle change of policies
2. Who triggers the mechanism
3. what is the desired behaviour after a change of policies
of the three (1) is has the highest a priority from m4d perspective, then (2).
What you describe as "creating a problem" is our current m4d behaviour - and addressing (1) would actually create the mechanism to externalize the change.

**Comment by ronenkat** [comment]
@roee88 looking again I suggest the following approach:
Add "PoliciesAppliedAfter" field to the spec, where the value can be either timestamp or generation.
The "PoliciesAppliedAfter" field complies with a definition of desired state and will enable reflecting the timestamp for when the policies were evaluated and put into effect. 
The m4dapp controller will compare "PoliciesAppliedAfter" with the a timestamp/generation in the status.
If the status value is greater than no changes are needed
If current timestamp/generation is smaller than PoliciesAppliedAfter" then its a future action, no need for changes yet (add a future requete) --- this address observability and grace periods.
Otherwise the policies could be re-evaluated and new blueprint would be generated.

**Reply by shlomitk1** [reply]
Changing the spec is less preferable in this case. 
-  An immediate reconcile occurs with generation changing, so the controller will try to reconcile even if an update is required for two weeks since today. (It would not know what was changed in the spec). 
-  It does not come from the user, so spec does not seem the right place.

Changing annotation seems to work better. The reconcile occurs, no change in generation, so the controller will check status of owned resources (was plotter status updated?) and also the new annotation filed. According to the update timestamp, it will either reconcile (immediate update required) or will requeue to trigger a future reconcile closer to the update time.

While we are focusing on policies,  this is relevant for other cases as well: a module is going to be removed/updated due to security reasons - reconstruct the blueprint when requested, the dataset is moving to a new location, etc.
Thus, we can add a general annotation, for example: `updateRequired: <timestamp>`
